Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp

```diff
@@ -137,6 +138,292 @@ void Client::ThreadFunc()
+        if ( minutes > 0 )
+        {
+            timeIntervalString.Format( "%um %us\n", minutes, seconds );
+        }
+        else
+        {
+            timeIntervalString.Format( "%us\n", seconds );
+        }
+    }
+    return shouldRetry;
+}
+
+// RetryWorkers
+//------------------------------------------------------------------------------
+void Client::RetryWorkers()
+{
+    if ( m_RetryingWorkers )
+    {
+        if ( m_SettingsWorkers.IsEmpty() )  // if dynamic workers
+        {
+            float timeSinceLastRetry = m_WorkerListRefreshTimer.GetElapsed();
+            if ( timeSinceLastRetry > m_WorkerListRefreshIntervalSec )
+            {
+                // reset timer here, since will be refreshing the worker list below
+                m_WorkerListRefreshTimer.Start();  // reset time
+                if ( m_WorkerListRefreshIntervalSec == 0 )
+                {
+                    m_WorkerListRefreshIntervalSec = 5;  // next retry is in >= 5 seconds
+                }
+                else
+                {
+                    // so we don't hit the network file share so much, use exponential backoff
+                    m_WorkerListRefreshIntervalSec *= 2;
+                }
+
+                Array< AString > previousWorkers(m_WorkerList);
+
+                m_WorkerList.Clear();
+                m_WorkerBrokerage.FindWorkers(
+                    m_ExcludedWorkers,
+                    m_WorkerList );
+
+                Array< AString > currentWorkers(m_WorkerList);
+                if ( currentWorkers.IsEmpty() )
+                {
+                    FLOG_WARN( "No matching workers found in '%s'", m_WorkerBrokerage.GetBrokerageRoot().Get() );
+                }
+
+                Array< AString > addedWorkers;
+                GetWorkerChanges( previousWorkers, currentWorkers, addedWorkers );
+                // don't remove any workers during refresh, since previous workers could be rebooting
+                Array< AString > removedWorkers;
+                UpdateServerList( removedWorkers, addedWorkers );
+
+                const size_t numPreviousWorkers( previousWorkers.GetSize() );
+                const size_t numCurrentWorkers( currentWorkers.GetSize() );
+                if ( numCurrentWorkers != numPreviousWorkers )
+                {
+                    OutputNumWorkers();
+                }
+
+                AStackString<> timeIntervalString;
+                AStackString<> timeLimitString;
+                if ( ShouldRetry( timeSinceLastRetry,
+                    m_WorkerListRefreshIntervalSec,
+                    m_WorkerListRefreshLimitSec,
+                    m_WorkerListRefreshElapsedTimeSec,
+                    timeIntervalString,
+                    timeLimitString ) )
+                {
+                    OUTPUT( "Will refresh worker list in %s, retry limit %s\n", timeIntervalString.Get(), timeLimitString.Get() );
```

Comment:
I don't think we want this output being emitted by default.

The DIST_INFO macro exists in the client to allow for output that will only be emitted with -distverbose is specified.